### PR TITLE
Feature/listener refactor

### PR DESCRIPTION
**Pull request by jagill**
@jjackson26 @swarajban This is refactor of the `recognizer.onresult` callback, which simplifies it a bit, as well as allowed me to find a bug.

The gist of it is:
1. Separate out the logic that parses the recognizer result event into a listener result event.
2. Instead of storing an interim result as the final entry of the `listener._results` object, it stores it in its own `listener._pendingResult` variable, which makes the logic for handling it much more comprehensible.

There are also some additions that allow recording events, as well as a FakeRecognizer helper class for testing, and new tests.  But these are secondary.

### CODE REVIEW

File: src/sdk/listener.js [hunk 1]
```diff
@@ -2,6 +2,36 @@
...
+  var ts = function(date) {
+    date = date || new Date();
+    return '[' +
+      pad(date.getHours(), 2) + ':' +
+      pad(date.getMinutes(), 2) + ':' +
+      pad(date.getSeconds(), 2) + '.' +
+      pad(date.getMilliseconds(), 3) +
+    ']';
+  };
+  // ZZZ END DEBUG STUFF
```
Comment (jjackson26) [inline src/sdk/listener.js]: Do we want to remove this debug code?

File: src/sdk/listener.js [hunk 2]
```diff
@@ -415,33 +449,35 @@
...
-      var results = listener._results;
-      var resultIndex = results.length - 1;
-      if (resultIndex >= 0) {
-        lastResult = results[resultIndex];
-        if (!lastResult.final) {
-          resultFinalized = lastResult.final = true;
-          lastResult.early = true;
-          listener.emit('result', lastResult, resultIndex, results, event);
-        }
+      if (listener._pendingResult) {
```
Comment (jjackson26) [inline src/sdk/listener.js]: Style: maybe change `function finalizeResult() {` to `var finalizeResult = function () {` for consistency.

File: src/sdk/listener.js [hunk 3]
```diff
@@ -496,22 +505,63 @@
...
+      var result = parseResult(event);
+
+      // If pendingResultFinalized, then we've made a non-final result
+      // into a final result and submitted it.  In this case, we want to
+      // ignore successive non-final results (because they are modifying
+      // something we've already finalized), and ignore final results (
+      // because they are duplicates) but when we get final results we
+      // want to start listening again.  This, when the result is final
+      // set resultFinalzed = false.
+      if (pendingResultFinalized && result.final) {
```
Comment (jjackson26) [inline src/sdk/listener.js]: Nice clear explanation.

File: src/test/spec/FakeRecognizer.js [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+      now = this._currentTimestamp();
+      jasmine.clock().tick(now - lastNow);
+      lastNow = now;
+      this._nextEvent();
+    }
+
+    jasmine.clock().uninstall();
+  };
+
+})();
```
Comment (jjackson26) [inline src/test/spec/FakeRecognizer.js]: Cool stuff

**Comment by jjackson26** [comment]
LGTM. It would be nice to have Swaraj see if he can easily make his way through the logic, as he's less acquainted with it.

File: src/sdk/listener.js [hunk 5]
```diff
@@ -496,22 +505,63 @@
       }
 
       // if we restarted, we'll need to add a space for some results
-      if (resultIndex >= 0 && !/^\s/.test(result.transcript.charAt(0))) {
+      if (listener._results.length > 0 && !/^\s/.test(result.transcript.charAt(0))) {
```
Comment (swarajban) [inline src/sdk/listener.js]: not relevant to this PR really, but why is the whitespace test a regex? Why can't it be `result.transcript.charAt(0) !== ' '`

File: src/sdk/listener.js (hunk 3 above)
Comment (swarajban) [inline src/sdk/listener.js]: +1

File: src/sdk/listener.js [hunk 6]
```diff
@@ -496,22 +505,63 @@
...
-      results[resultIndex] = result;
-      if (!result.final) {
+      if (result.final) {
+        window.clearTimeout(earlyFinalResultTimeout);
+        earlyFinalResultTimeout = null;
+        listener._pendingResult = null;
+        listener._results.push(result);
+      } else {
+        // Remember this in case we need to finalize it.
+        listener._pendingResult = result;
```
Comment (swarajban) [inline src/sdk/listener.js]: I guess I forgot, but we only store one _pendingResult?

File: src/sdk/listener.js [hunk 7]
```diff
@@ -415,33 +449,35 @@
...
+        pendingResultFinalized = true;
       }
     }
 
-    recognizer.onresult = function(event) {
-      // If the debug flag is set, capture this for debugging.
-      if (listener._captureRaw) {
-        listener._rawResults.push(event);
-      }
+    var registerEvent = function (name, e) {
```
Comment (swarajban) [inline src/sdk/listener.js]: Can you add a comment describing what this function does?

Comment (swarajban) [inline src/sdk/listener.js]: Ah, just for unit testing ya?

File: src/sdk/main.js [hunk 8]
```diff
@@ -859,7 +859,6 @@ var MM = ( function (window, ajax, Faye) {
               MM.activeUser.sessions.post(
                 session,
                 function onSessionCreate(result) {
-                  console.log('Create session results:', result);
```
Comment (swarajban) [inline src/sdk/main.js]: Cool, are there any more of these lying around?

File: src/test/spec/FakeRecognizer.js [hunk 9]
```diff
@@ -0,0 +1,118 @@
+/* global spyOn, jasmine */
+
+;(function () {
+
+  var FakeRecognizer = window.FakeRecognizer = function () {
```
Comment (swarajban) [inline src/test/spec/FakeRecognizer.js]: Can you add a brief description of this class in comments (I know *we* all know it's a fake webkit recognizer, but let's write it down here)

File: src/test/spec/FakeRecognizer.js [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+    spyOn(this, 'start').and.callThrough();
+    spyOn(this, 'stop').and.callThrough();
+    spyOn(this, 'abort').and.callThrough();
+
+    this._isStarting = false;
+  };
+
+  FakeRecognizer.prototype.start = function () {
+    if (this._isStarting) {
+      // var e = new DOMError("InvalidStateError");
```
Comment (swarajban) [inline src/test/spec/FakeRecognizer.js]: what is this?

File: src/test/spec/FakeRecognizer.js [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+    this._isStarting = true;
+    var self = this;
+    setTimeout(function () {
+      self.onstart && self.onstart();
+    }, 10);
+  };
+
+  FakeRecognizer.prototype._end = function () {
+    this._isStarting = false;
+    this.onend && this.onend();
```
Comment (swarajban) [inline src/test/spec/FakeRecognizer.js]: Just supports a single event subscriber ya?

File: src/test/spec/FakeRecognizer.js [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+
+  FakeRecognizer.prototype.abort = function () {
+    var self = this;
+    setTimeout(function () {
+      self._abort();
+    }, 10);
+  };
+
+  FakeRecognizer.prototype._currentTimestamp = function () {
+    if (this._eventIndex >= this._events.length) {
```
Comment (swarajban) [inline src/test/spec/FakeRecognizer.js]: nit: can't you just use a `this._currentEvent === null` check?

File: src/test/spec/FakeRecognizer.js (hunk 4 above)
Comment (swarajban) [inline src/test/spec/FakeRecognizer.js]: Ya, cool man

File: src/test/spec/ListenerPlaybackTest.js [hunk 13]
```diff
@@ -0,0 +1,262 @@
...
+    var listener, recognizer, results;
+
+    beforeEach(function () {
+      MM.support = { speechRecognition: true };
+      recognizer = new FakeRecognizer();
+
+      listener = new MM.Listener({interimResults: true});
+      listener._recognizer = recognizer;
+      listener._initializeRecognizer(recognizer);
+      recognizer._setEvents(Playbacks.pizza1);
```
Comment (swarajban) [inline src/test/spec/ListenerPlaybackTest.js]: haha what a name

**Comment by swarajban** [comment]
This looks a lot better than before. Really like the comments & the recording functionality

File: src/sdk/listener.js (hunk 5 above)
Comment (jagill) [inline src/sdk/listener.js]: Good question -- I'm still a little unsure about when we need this;  I'd love some more tests showing the cases.

File: src/sdk/listener.js (hunk 6 above)
Comment (jagill) [inline src/sdk/listener.js]: The flow is that we get N interimResults, usually followed by a a final result (that's the completion of those interim results).  The interim results are like ['hi', 'hi how', 'hi, how are', 'hi, how are you'], so we only want the last (which has the highest confidence).  Thus, at any given time, there is only one relevant interim result, which is the latest.

File: src/sdk/main.js (hunk 8 above)
Comment (jagill) [inline src/sdk/main.js]: Not that I saw.

File: src/test/spec/FakeRecognizer.js (hunk 11 above)
Comment (jagill) [inline src/test/spec/FakeRecognizer.js]: Yes, the (real) recognizer just has `onend` etc hooks that you set; it's unfortunately not evented.

File: src/test/spec/ListenerPlaybackTest.js (hunk 13 above)
Comment (jagill) [inline src/test/spec/ListenerPlaybackTest.js]: Yeah, i originally was going to name transcripts by the content, but then i started naming them by their function.  I didn't go back to correct the old ones, dunno if I should!

**Comment by jagill** [comment]
Ok, fixed code review comments.  I've merged a squashed version (there was a large file I added then deleted) and will push that.  Don't merge this PR; i'll close it in a bit.

**Pull request closed without merging**